cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,364 @@
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve list of firewalls from DO API: %v", err)
+	}
+	return fw, nil
+}
+
+// Set applies the given inbound rules to the public access firewall when the current rules and target rules differ.
+// Set will also reconcile away any changes to the outbound rules, name and tags.
+func (fm *firewallManagerOp) Set(ctx context.Context, fr *godo.FirewallRequest) error {
+	targetFirewall := fm.fwCache.getCachedFirewall()
+
+	if targetFirewall != nil {
+		equal, _ := isEqual(targetFirewall, fr)
+		if equal {
+			// A locally cached firewall with matching rules, correct name and tags means there is nothing to update.
+			return nil
+		}
+		// A locally cached firewall exists, but the inbound rules don't match the expected
+		// service inbound rules, outbound rules, name or tags. So we need to use the locally
+		// cached firewall ID to attempt to update the firewall APIs representation of the
+		// firewall with the new rules.
+		currentFirewall, resp, err := fm.client.Firewalls.Update(ctx, targetFirewall.ID, fr)
+		if err != nil {
+			if resp == nil || resp.StatusCode != http.StatusNotFound {
+				return fmt.Errorf("could not update firewall: %v", err)
+			}
+			currentFirewall, err = fm.createFirewall(ctx, fr)
+			if err != nil {
+				return fmt.Errorf("could not create firewall: %v", err)
+			}
+			klog.Info("successfully created firewall")
+		}
+		fm.fwCache.updateCache(currentFirewall)
+		return nil
+	}
+
+	// Check if the target firewall ID exists but in the case that CCM first starts up and the
+	// firewall ID does not exist yet, we have to retrieve one from the API.
+	if targetFirewall == nil {
+		currentFirewall, err := fm.Get(ctx)
+		if err != nil {
+			return err
+		}
+		if currentFirewall == nil {
+			klog.Info("an existing firewall not found, we need to create one")
+			currentFirewall, err = fm.createFirewall(ctx, fr)
+			if err != nil {
+				return err
+			}
+			klog.Info("successfully created firewall")
+		}
+		fm.fwCache.updateCache(currentFirewall)
+	}
+	return nil
+}
+
+func isEqual(targetFirewall *godo.Firewall, fr *godo.FirewallRequest) (bool, []string) {
+	var unequalParts []string
+	if targetFirewall.Name != fr.Name {
+		unequalParts = append(unequalParts, "name")
+	}
+	if !cmp.Equal(targetFirewall.InboundRules, fr.InboundRules) {
+		unequalParts = append(unequalParts, "inboundRules")
+	}
+	if !cmp.Equal(targetFirewall.OutboundRules, fr.OutboundRules) {
+		unequalParts = append(unequalParts, "outboundRules")
+	}
+	if !cmp.Equal(targetFirewall.Tags, fr.Tags) {
+		unequalParts = append(unequalParts, "tags")
+	}
+	return len(unequalParts) == 0, unequalParts
+}
+
+func (fm *firewallManagerOp) createFirewall(ctx context.Context, fr *godo.FirewallRequest) (*godo.Firewall, error) {
+	currentFirewall, _, err := fm.client.Firewalls.Create(ctx, fr)
+	return currentFirewall, err
+}
+
+// create a firewall request that has the correct rules, name and tag
```

Comment:
GoDocs for variables, functions, and methods should always start with the name of the documentee; that is, this comment should start like

```go
// createReconciledFirewallRequest [...]
```